api/coreservice.go

```diff
@@ -311,16 +326,38 @@ func (core *coreService) Account(addr address.Address) (*iotextypes.AccountMeta,
 	if addrStr == address.RewardingPoolAddr || addrStr == address.StakingBucketPoolAddr {
 		return core.getProtocolAccount(ctx, addrStr)
 	}
-	span.AddEvent("accountutil.AccountStateWithHeight")
 	ctx = genesis.WithGenesisContext(ctx, core.bc.Genesis())
-	state, tipHeight, err := accountutil.AccountStateWithHeight(ctx, core.sf, addr)
+	return core.acccount(ctx, core.bc.TipHeight(), false, core.sf, addr)
+}
+
+func (core *coreService) acccount(ctx context.Context, height uint64, archive bool, sr protocol.StateReader, addr address.Address) (*iotextypes.AccountMeta, *iotextypes.BlockIdentifier, error) {
+	span := tracer.SpanFromContext(ctx)
+	span.AddEvent("accountutil.AccountStateWithHeight")
+	state, height, err := accountutil.AccountStateWithHeight(ctx, sr, addr)
 	if err != nil {
 		return nil, nil, status.Error(codes.NotFound, err.Error())
 	}
 	span.AddEvent("ap.GetPendingNonce")
-	pendingNonce, err := core.ap.GetPendingNonce(addrStr)
-	if err != nil {
-		return nil, nil, status.Error(codes.Internal, err.Error())
+	var (
+		addrStr      = addr.String()
+		pendingNonce uint64
+	)
+	if archive {
+		state, err := accountutil.AccountState(ctx, sr, addr)
+		if err != nil {
+			return nil, nil, status.Error(codes.NotFound, err.Error())
+		}
+		g := core.bc.Genesis()
+		if g.IsSumatra(height) {
+			pendingNonce = state.PendingNonceConsideringFreshAccount()
+		} else {
+			pendingNonce = state.PendingNonce()
+		}
+	} else {
+		pendingNonce, err = core.ap.GetPendingNonce(addrStr)
```

Comment:
Recommend to move 'nonce from the actpool` to the Account method, and removing the 'archive bool' parameter would make it clearer.

Reply:
updated